### PR TITLE
Improve fleet tooltip for fighters

### PR DESCRIPTION
**Pull request by dbenage-cx**
Provides two tooltip modes for a ships fighter icon in fleet window.
The first tooltip displays after the normal delay, with a summary view.
The second replaces the first after a longer delay, with a more detailed accounting.

[Forum discussion](http://www.freeorion.org/forum/viewtopic.php?f=6&t=10342)

### CODE REVIEW

File: UI/MeterBrowseWnd.cpp [hunk 1]
```diff
@@ -580,4 +585,360 @@ void ShipDamageBrowseWnd::UpdateEffectLabelsAndValues(GG::Y& top) {
...
+    m_effect_labels_and_values.push_back(std::make_pair(bay_summary, hangar_summary));
+
+    top += m_row_height;
+
+    std::string fighter_damage_text = ColourWrappedtext(DoubleToString(fighter_damage, 3, false), DAMAGE_COLOR);
+
+    if (!m_show_all_bouts) {
+        // Show damage for first wave (2nd combat round)
+        std::pair<int, float> first_wave = FightersDamageForBout(bay_total_capacity, hangar_current_fighters,
+                                                                    fighter_damage, 2);
```
Comment (geoffthemedio) [inline UI/MeterBrowseWnd.cpp]: indenting

File: UI/MeterBrowseWnd.cpp [hunk 2]
```diff
@@ -580,4 +585,360 @@ void ShipDamageBrowseWnd::UpdateEffectLabelsAndValues(GG::Y& top) {
...
+        top += m_row_height;
+    } else {
+        // add labels for bay parts
+        for (std::multimap<std::string, int>::value_type& bay_part : bay_parts) {
+            const std::string& part_name = UserString(bay_part.first);
+            int fighter_total = bay_part.second;
+            m_bay_list->Insert(new ShipFightersBrowseRow(part_name, fighter_total));
+        }
+
+        // add a label for each hangar part, distributing the total fighters equally
```
Comment (geoffthemedio) [inline UI/MeterBrowseWnd.cpp]: could instead have "hangar_part x 6" and just the sum of fighters in that hangar part type...

File: UI/MeterBrowseWnd.cpp [hunk 3]
```diff
@@ -580,4 +585,360 @@ void ShipDamageBrowseWnd::UpdateEffectLabelsAndValues(GG::Y& top) {
...
+            // top heavy when current fighters do not distribute equally
+            int cur_fighters = std::min(unaccounted_current_fighters, std::min(max_fighters,
+                boost::math::iround(static_cast<float>(hangar_current_fighters) / hangar_part.second)));
+            unaccounted_current_fighters -= cur_fighters;
+            unaccounted_total_fighters -= max_fighters;
+            m_hangar_list->Insert(new ShipFightersBrowseRow(hangar_name, cur_fighters, max_fighters));
+        }
+        // validate that all fighters were accounted for
+        if (unaccounted_current_fighters != 0 || unaccounted_total_fighters != 0) {
+            DebugLogger() << "Display of number of fighters mismatched.  Current: "
```
Comment (geoffthemedio) [inline UI/MeterBrowseWnd.cpp]: Should the be ErrorLogger() ?

File: UI/MeterBrowseWnd.cpp [hunk 4]
```diff
@@ -580,4 +585,360 @@ void ShipDamageBrowseWnd::UpdateEffectLabelsAndValues(GG::Y& top) {
...
+        m_bay_list->Resize(GG::Pt(ROW_WIDTH, LIST_HEIGHT));
+        m_hangar_list->MoveTo(GG::Pt(ROW_WIDTH + EDGE_PAD, top));
+        m_hangar_list->Resize(GG::Pt(ROW_WIDTH, LIST_HEIGHT));
+
+        top += LIST_HEIGHT;
+
+        // Damage summary labels
+        float total_dmg = FightersDamageForBout(bay_total_capacity, hangar_current_fighters, fighter_damage).second;
+        std::string detail_avg_text = boost::io::str(FlexibleFormat(UserString("SHIP_FIGHTERS_DAMAGE_AVERAGE"))
+            % ColourWrappedtext(DoubleToString(total_dmg / NUM_COMBAT_BOUTS, 3, false), DAMAGE_COLOR));
```
Comment (geoffthemedio) [inline UI/MeterBrowseWnd.cpp]: indenting in this case is inconsistent throughout these additions. should perhaps be consistent with indendenting 4 speces or to the same position as FlexibleFormat

File: UI/MeterBrowseWnd.h [hunk 5]
```diff
@@ -70,4 +70,22 @@ class ShipDamageBrowseWnd : public MeterBrowseWnd {
 
 };
 
+class ShipFightersBrowseWnd : public MeterBrowseWnd {
+public:
+    ShipFightersBrowseWnd(int object_id, MeterType primary_meter_type, bool show_all_bouts = false);
+    static const int NUM_COMBAT_BOUTS = 3;
```
Comment (geoffthemedio) [inline UI/MeterBrowseWnd.h]: as discussed on forum thread, don't duplicate this definition here...

**Review by geoffthemedio — CHANGES_REQUESTED** [review]
a few more comments before various code blocks would also be nice.

**Comment by Vezzra** [comment]
The link to the forum discussion is somehow broken... when clicking it, I get a forum information message "No post mode specified".

Edit: Fixed it -Geoff /Edit

**Comment by geoffthemedio** [comment]
Builds and seems to work on Windows. Default long tooltip time (5 s) is too long, though.

File: UI/MeterBrowseWnd.cpp (hunk 1 above)
Comment (dbenage-cx) [reply UI/MeterBrowseWnd.cpp]: Fixed

File: UI/MeterBrowseWnd.cpp (hunk 2 above)
Comment (dbenage-cx) [reply UI/MeterBrowseWnd.cpp]: Fixed, made bays consistent with this change

File: UI/MeterBrowseWnd.cpp (hunk 4 above)
Comment (dbenage-cx) [reply UI/MeterBrowseWnd.cpp]: Fixed

File: UI/MeterBrowseWnd.h (hunk 5 above)
Comment (dbenage-cx) [reply UI/MeterBrowseWnd.h]: Pending resolution of PR #1181

**Comment by dbenage-cx** [comment]
Rebased

File: UI/MeterBrowseWnd.cpp [hunk 6]
```diff
@@ -580,4 +592,386 @@ void ShipDamageBrowseWnd::UpdateEffectLabelsAndValues(GG::Y& top) {
...
+    class ShipFightersBrowseRow : public GG::ListBox::Row {
+    public:
+        /** Ctor
+         * @param [in] label text for the descriptive label, aligned to right
+         * @param [in] qty quantity of source effect in @p label, not displayed when less or equal to 1.
+         * @param [in] value value for label positioned to right of @p qty, colored if @p base_value is unspecified
+         * @param [in] base_value optional; If greater than 0.0f: the value label is formatted to "value of base_value"
+         */
+        ShipFightersBrowseRow(const std::string& label, int qty, double value, double base_value = 0.0f) :
+                GG::ListBox::Row(FighterBrowseListWidth(), MeterBrowseRowHeight(), "")
```
Comment (geoffthemedio) [inline UI/MeterBrowseWnd.cpp]: indent step is 4 not 8 spaces

Comment (dbenage-cx) [reply UI/MeterBrowseWnd.cpp]: TY, fixed.

**Review by geoffthemedio — APPROVED** [review]

**Comment by geoffthemedio** [comment]
fighter_tooltip_annotated

A couple quirks...

How are there 4 fighters able to shoot in round 2 if there are just 3 launch bays, and why is the number of fighters the same in both rounds?

The layout for listing / counting the launch bays is odd.

**Comment by MatGB** [comment]
Launch bays can each launch 2 fighters, so 3 can launch up to 6 in one turn, that all looks fine to me. 

I expect at some point to have bigger or better launch bays but for now haven't touched your numbers.

**Comment by dbenage-cx** [comment]
Substantial change to address layout and style, and provide more info on a round.

fighter_tt8

**Comment by Vezzra** [comment]
Looks ok to me... as far as I am concerned, I guess this can be merged. To get a feeling how useful that really is more extensive playtesting will be required anyway.

**Comment by dbenage-cx** [comment]
@geoffthemedio Thank you for reviews

**Pull request merged**